### DISCUSSION TITLE
codec can't encode character '\u266a'

### DISCUSSION THREAD

**Discussion by turnkit** [openai/whisper#729] [General]
When I run Whisper in a batch model on the large model it produces a musical symbol and fails. 

Seems to work fine though in a normal command line just not in a shell as below. Odd?

Any suggestions?

> UnicodeEncodeError: 'charmap' codec can't encode character '\u266a' in position 27: character maps to <undefined>

here's the script:
for f in *.mp4
  do echo 'Start processing: ' "${f}" $(date)
  # whisper  --language en --model large --device cuda --verbose True --condition_on_previous_text False --temperature_increment_on_fallback None --output_dir whisp_out_large "${f}"
  whisper --model large --language en --device cuda --verbose True --output_dir whisp_out_large "${f}"
  mv "${f}" "./_transcription_done_large/${f}"
  echo 'Stop processing: ' "${f}" $(date)
  done

Here's the full output:

> 
>   
> processing:  2021-1212-CCF-LiveStream-WorshipService.mp4 Mon Dec 19 17:20:52 CST 2022
> Traceback (most recent call last):
>   File "C:\Users\turnk\anaconda3\lib\runpy.py", line 197, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "C:\Users\turnk\anaconda3\lib\runpy.py", line 87, in _run_code
>     exec(code, run_globals)
>   File "C:\Users\turnk\anaconda3\Scripts\whisper.exe\__main__.py", line 7, in <module>
>   File "C:\Users\turnk\anaconda3\lib\site-packages\whisper\transcribe.py", line 307, in cli
>     result = transcribe(model, audio_path, temperature=temperature, **args)
>   File "C:\Users\turnk\anaconda3\lib\site-packages\whisper\transcribe.py", line 207, in transcribe
>     add_segment(
>   File "C:\Users\turnk\anaconda3\lib\site-packages\whisper\transcribe.py", line 168, in add_segment
>     print(f"[{format_timestamp(start)} --> {format_timestamp(end)}] {text}")
>   File "C:\Users\turnk\anaconda3\lib\encodings\cp1252.py", line 19, in encode
>     return codecs.charmap_encode(input,self.errors,encoding_table)[0]
> UnicodeEncodeError: 'charmap' codec can't encode character '\u266a' in position 27: character maps to <undefined>
> <Managed Device 0>
> Traceback (most recent call last):
>

**Comment by turnkit** [comment]
Ah - weird. Fixed the issue.

I was running on Windows in Anaconda shell.  Somehow when it launches a new window for a shell it doesn't handle the extended uni characters well.  I got around the problem by just typing 'bash' in the Anaconda shell and then launching the little script with bash so that all the whisper run stayed in that same environment. This is the "fun" of running Whisper under WSL2 in Windows 11 I guess.

**Comment by mihadoo** [comment]
The problem is the following line (under Windows 11):

c:\Python311\Lib\site-packages\whisper\transcribe.py line 339:

                    print(make_safe(line))

If I replace it to the following then it does not crash anymore but it also does not produce the correct output:

                    print(make_safe(line).encode('utf8'))

Maybe someone with enough python knowledge can correct this. A possible solution is described here:

https://stackoverflow.com/questions/10569438/how-to-print-unicode-character-in-python

In any case the output must be UTF-8 for TXT and JSON VTT and other files so that all languages will work correctly.

**Reply by magnumquest** [reply]
This solves the issue:
Edit the file /modules/utils.py

And change the function add_to_textfile to:

```
def add_to_textfile(text_file, text):
  text = text.encode('utf-8')
  with open(text_file, 'a', encoding='utf-8') as f:
    f.write(text.decode('utf-8'))
```

**Comment by Sparviero-Sughero** [comment]
add_to_textfile() function no longer exists in utils.py

**Comment by fbergmann** [comment]
I've encountered the same error. To resolve unicode erros in `make_save(line)` I've modified how to retrieve the locale that was currently used. This resolved it on windows

<https://github.com/fbergmann/whisper>

should i make a PR?

```diff
diff --git a/whisper/utils.py b/whisper/utils.py
index 13792f7..6f3d730 100644
--- a/whisper/utils.py
+++ b/whisper/utils.py
@@ -1,12 +1,12 @@
 import json
 import os
 import re
-import sys
+
 import zlib
 from typing import Callable, List, Optional, TextIO
+import locale

-system_encoding = sys.getdefaultencoding()
-
+system_encoding = locale.getpreferredencoding(False)
 if system_encoding != "utf-8":

     def make_safe(string):
```
